### PR TITLE
Adds `PostWallets` and `PatchWallets`.

### PR DESCRIPTION
**Pull request by szilardszaloki**
Resolves https://github.com/brave/brave-browser/issues/25484.
Security/privacy review: https://github.com/brave/security/issues/1034.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [x] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [x] New files have MPL-2.0 license header
- [x] Adequate test coverage exists to prevent regressions
- [x] Major classes, functions and non-trivial code blocks are well-commented
- [x] Changes in component dependencies are properly reflected in `gn`
- [x] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [x] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Make sure wallet creation works as before.
Please also spot check test cases from [14773](https://github.com/brave/brave-core/pull/14773)'s test plan.

### CODE REVIEW

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/request_for.h
Comment (zenparsing) [inline vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/request_for.h:41]: This should probably be run in a future turn, in order to avoid reentrancy surprises for callers. The basic pattern is:

```cpp
base::SequencedTaskRunnerHandle::Get()->PostTask(
    FROM_HERE, 
    base::BindOnce(std::move(callback), Endpoint::Error::FAILED_TO_CREATE_REQUEST));
```

File: vendor/bat-native-ledger/src/bat/ledger/internal/bitflyer/bitflyer_authorization.cc
Comment (zenparsing) [inline vendor/bat-native-ledger/src/bat/ledger/internal/bitflyer/bitflyer_authorization.cc:174]: Nit: I think this [style rule](https://chromium.googlesource.com/chromium/src/+/master/styleguide/c++/c++-dos-and-donts.md#variable-initialization) suggests using parens for constructors.

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/error_types.mojom
Comment (zenparsing) [inline vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/error_types.mojom:1]: I think this file needs to be in the same folder as `ledger.mojom`.

Do we need the nested namespace here? I think `bat-native-ledger` in general went overboard with nested namespacing; all these names should be unique anyway, and it just creates more typing. I think there's a Chromium [style guide](https://chromium.googlesource.com/chromium/src.git/+/HEAD/docs/security/mojo.md#naming-conventions) for this as well.

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/request_for.h
Comment (zenparsing) [inline vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/request_for.h:42]: I'm wondering if there's another option besides a generic "failed to create request" error that's duplicated across all result enums.

The other option that I can think of is to have the RequestBuilder methods go back to returning `base::expected`, but with an `Error` as the error type.

```cpp
base::expected<std::string, MyEndpointError> Content() override {
  return base::unexpected(MyEndpointError::kInvalidRewardsWallet);
}
```

But then `RequestBuilder` would also need to be CRTP.

It seems like it's fine as-is for now though.

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/error_types.mojom
Comment (szilardszaloki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/error_types.mojom:4]: Ah, right — also found it [here](https://google.github.io/styleguide/cppguide.html#Enumerator_Names).

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/patch_wallets/patch_wallets.cc
Comment (szilardszaloki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/patch_wallets/patch_wallets.cc:147]: Makes sense, adding a comment there.

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/patch_wallets/patch_wallets.cc
Comment (szilardszaloki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/patch_wallets/patch_wallets.cc:151]: Good idea.

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/request_for.h
Comment (szilardszaloki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/request_for.h:42]: We'd end up having several interface classes (`RequestBuilder<>`, one per endpoint) instead of just one (`RequestBuilder`), since e.g. `base::expected<std::string, MyEndpointError1>` and `base::expected<std::string, MyEndpointError2>` are neither the same type, nor covariant types. That's just a side effect, but we'd also run into a "chicken or the egg" problem:
```
template <typename Endpoint>
class RequestBuilder {
 public:
// ...
  virtual base::expected<std::string, /* >>> */ typename Endpoint::Error /* <<< */> Content();
// ...
};
```
At that point `Endpoint` is an incomplete type, so you can't use its nested typedef.
  There would be workarounds (e.g. passing `Error` as `RequestBuilder<>`'s template parameter, using a trait class, building a comple-time map and do a lookup while `RequestBuilder<>` is instantiated, etc.), but neither is optimal, unfortunately.

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/request_for.h
Comment (szilardszaloki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/request_for.h:41]: Hmm, would you mind elaborating on this a bit?

File: vendor/bat-native-ledger/src/bat/ledger/internal/bitflyer/bitflyer_authorization.cc
Comment (szilardszaloki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/bitflyer/bitflyer_authorization.cc:174]: Dang, Chromium seems to be a bit overconcerned about variable initialization...
In contrast, [here](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es23-prefer-the--initializer-syntax)'s what the C++ Core Guidelines has to say about it — amending regardless.

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/error_types.mojom
Comment (szilardszaloki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/error_types.mojom:1]: I think it's a good idea to keep `error_types.mojom` as close to where it belongs as possible, as opposed to putting it in `vendor/bat-native-ledger/include/bat/ledger/public/interfaces`.

Same is true for the error types in the nested `endpoints` namespace.
Also, the compile-time machinery that checks if a certain `enum class` (and its `FAILED_TO_CREATE_REQUEST` enumerator) exists relies on some basic code generation/macro expansion, e.g.:
```
RESULT_FOR(PatchWallets, void);
```
I wanted this to be put right above the declaration of the endpoint class (`PatchWallets`, in this case). `RESULT_FOR()` opens up the `mojom` namespace and forward declares the expected `enum class` as part of the check. If the error types are defined in `ledger::mojom`, I have to start the macro definition with a closing brace (since we're in the `endpoints` namespace), which I didn't like (and neither did the linter 🙂).

By the way: we wouldn't have this problem, if we didn't adhere to Chromium's naming convention (that is, to put every Mojo type in a nested `mojom` namespace), which is just a guideline to avoid name collisions (and which they even proposed to get rid of in 2019). I'm just not sure if we really want to adopt this otherwise completely useless convention?

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/patch_wallets/patch_wallets.cc
Comment (szilardszaloki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/patch_wallets/patch_wallets.cc:147]: Come to think of it, this is not a `RequestBuilder`-specific thing, as it's not `RequestBuilder` who relies on the ordering of calls to `Content()` and `Headers()`, but the endpoint implementations themselves.

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/request_for.h
Comment (szilardszaloki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/request_for.h:41]: Linking our [conversation](https://github.com/brave/brave-core/pull/15108#discussion_r972402047) here from your PR.

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/patch_wallets/patch_wallets.cc
Comment (szilardszaloki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/patch_wallets/patch_wallets.cc:147]: Actually, adjusting `RequestBuilder::Request()` a bit:
```
// ...

const auto content = Content();
if (!content) {
  return absl::nullopt;
}

auto headers = Headers(*content);
if (!headers) {
  return absl::nullopt;
}

// ...
```
so that endpoint implementations don't have to rely on the ordering of these two calls.

Comment (zenparsing) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/patch_wallets/patch_wallets.cc:147]: This makes the ordering explicit rather than implicit, which is better (and self-documenting).

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/error_types.mojom
Comment (zenparsing) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/error_types.mojom:1]: Is there a way to do this without the macro? I would prefer that we not use macros to declare things if possible. Can the endpoint just define an alias for the Error type?

Comment (szilardszaloki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/error_types.mojom:1]: Yeah, I personally hate macros, but I can't think of a cleaner way of making sure:
- the error type in `error_types.mojom` is called `[Endpoint]Error`
- `[Endpoint]Error` has the enumerator `kFailedToCreateRequest`
- the endpoint implementor doesn't have to provide a forward declaration of their endpoint class, and an explicit specialization of the traits class:
  ```
  class Endpoint;

  template <>
  struct ResultFor<Endpoint> {
    using Result = base::expected<value_type, EndpointError>;
  };
  ```

Comment (zenparsing) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/error_types.mojom:1]: Can we drop `ResultFor<X>` and just do something like:

```cpp
class MyEndpoint : public RequestBuilder, public ResponseHandler<MyEndpoint> {
 public:
  using Result = base::expected<std::string, mojom::MyEndpointError>;
};
```

The compiler will catch the first two items, IIUC.

Or I suppose it could be

```cpp
class MyEndpoint : public RequestBuilder, public ResponseHandler<MyEndpoint> {
 public:
  using Value = std::string;
  using Error = mojom::MyEndpointError;
};
```

And still let `ResponseHandler` define the `Result` type.

Comment (szilardszaloki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/error_types.mojom:1]: Here:
```cpp
class MyEndpoint : public RequestBuilder, public ResponseHandler<MyEndpoint> {
 public:
  using Result = base::expected<std::string, mojom::MyEndpointError>;
};
```
it's up to the implementor what they want the typedef to be called (e.g. `NotResult`), but certain parts of the Request/Response API (e.g. `RequestFor<>`) rely on this typedef being called `Result` (that's why I want it to be defined in `ResponseHandler<>`). As for the first two items, there's nothing that would prevent the implementor from naming the error type however they like, also, the presence of `kFailedToCreateRequest` is not enforced (other than a cryptic compilation error).

Regarding this one:
```cpp
class MyEndpoint : public RequestBuilder, public ResponseHandler<MyEndpoint> {
 public:
  using Value = std::string;
  using Error = mojom::MyEndpointError;
};
```
It doesn't work because of the reasons outlined [here](https://github.com/brave/brave-core/pull/15101#discussion_r972252048).

Comment (zenparsing) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/error_types.mojom:1]: Get rid of `ResponseHandler` then? It feels like the design space is really tightly constrained.

Comment (szilardszaloki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/error_types.mojom:1]: The point of this design would be to enforce endpoint implementors to follow certain guidelines (and to not have any means of diverging from them), while requiring writing the least amount of code to define an endpoint. As a consequence, it doesn't leave too much design space, but this is to end up with endpoint implementations that resemble each other as much as possible. This adheres to Scott Meyers' `Make interfaces easy to use correctly and hard to use incorrectly.` advice. 🙂 Removing `ResponseHandler<>` would pretty much defeat the whole purpose.

File: components/brave_rewards/resources/page/BUILD.gn
Comment (zenparsing) [inline components/brave_rewards/resources/page/BUILD.gn:10]: I'm not completely sure about how dependencies are supposed to be laid out, but I think the intention is that other parts of the codebase shouldn't reach into `bat/ledger/internal`. I'm fairly certain that this needs to be in `public`.

File: vendor/bat-native-ledger/src/bat/ledger/internal/wallet/wallet_create.h
Comment (zenparsing) [inline vendor/bat-native-ledger/src/bat/ledger/internal/wallet/wallet_create.h:23]: Generally, the style is to have the callback go after the args? We should probably stick with that convention unless we plan on moving the codebase to that pattern.

Comment (szilardszaloki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/wallet/wallet_create.h:23]: Yeah, good point, done.

File: components/brave_rewards/resources/page/BUILD.gn
Comment (szilardszaloki) [reply components/brave_rewards/resources/page/BUILD.gn:10]: Put `error_types.mojom` in `vendor/bat-native-ledger/include/bat/ledger/public/interfaces` and renamed it to `ledger_endpoints.mojom`. (Not triggering JavaScript generation here for now.)

**Review by zenparsing — APPROVED** [review]

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/error_types.mojom
Comment (szilardszaloki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/error_types.mojom]: Fixed in 0a230cd08129036ba2e22092cb89335a75897b35 (as per our Slack convo).

Comment (bsclifton) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/error_types.mojom]: I personally (for whatever that is worth 😂) preferred the `RESULT_FOR` macro, since it encapsulates things a bit more (ex: makes its own enum, defines the class). But this edit looks OK too (fairly minimal since we only have 3 implementations here)

I definitely do understand how the macro can feel dirty; it reminds me of MFC a little bit (on older MSVC++ projects). But the usage was concise 👍  Glad you all were able to sort it out though 😄

**Review by kdenhartog — COMMENTED** [review]
I'm pretty sure these `base::JSONReader::Read()` calls are safe, but I need to double check it. I can see they will use rust if the flag is set for it, and I believe if it's not it will spawn a separate thread to sandbox the parser when it's checking the JSON.

I'll resolve those if they're not an issue. Few other small things as well, but nothing major here.

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/patch_wallets/patch_wallets.cc
Comment (szilardszaloki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/patch_wallets/patch_wallets.cc:26]: Could you expound on this a bit more?

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/post_wallets/post_wallets.cc
Comment (szilardszaloki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/post_wallets/post_wallets.cc:94]: That's by design, as `POST /v4/wallets` is the one that returns the Rewards payment ID to the client. [Here](https://docs.google.com/document/d/11W-6HV47aUt6sXMbt10z0x55ShtM27VwrOsquF6EzQs/edit#heading=h.vfxn9ps2qv2)'s the spec on wallet creation from `v3`.

File: vendor/bat-native-ledger/src/bat/ledger/internal/wallet/wallet_create.cc
Comment (szilardszaloki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/wallet/wallet_create.cc:28]: I can see how this can be confusing, so renaming it back to `Start()`. `WalletCreate` and `WalletRecover` are unnecessary abstractions. Instead, `Wallet` itself should have functions called `Create()` and `Recover()`. Will address these in a follow-up PR.

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/patch_wallets/patch_wallets.cc
Comment (kdenhartog) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/patch_wallets/patch_wallets.cc:26]: It's a task for myself. I've seen the JSON Sanitizer used more frequently, but essentially this JSON is an untrusted input so it needs to be parsed while following the rule of 2.

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/patch_wallets/patch_wallets.cc
Comment (szilardszaloki) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/patch_wallets/patch_wallets.cc:126]: `BLOG` is the ledger's logger facility, `VLOG` is Chromium's.

**Comment by szilardszaloki** [comment]
Hi @kdenhartog, 🙂 
This PR is currently blocked by security review. I was wondering if we have any security-related issues here? If not, then we might want to proceed with merging, as this PR is resolving a `P1` (and is used as the basis of @zenparsing's work, which is a `P1` itself).
Thanks!
cc: @Miyayes

**Comment by szilardszaloki** [comment]
As far as `base::JSONReader::Read()` and `base::StringPrintf()`, these are used in almost every endpoint implementation (and elsewhere). If either turns out to be unsafe, we should fix it everywhere, preferably in a follow-up, in my opinion.

**Comment by szilardszaloki** [comment]
Created https://github.com/brave/brave-browser/issues/25628 as per our Slack conversation with @kdenhartog.

File: vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/patch_wallets/patch_wallets.cc
Comment (kdenhartog) [reply vendor/bat-native-ledger/src/bat/ledger/internal/endpoints/patch_wallets/patch_wallets.cc:126]: Thanks, resolving this one.

**Review by kdenhartog — APPROVED** [review]
Approving for security to get this work unblocked for other ongoing work and we'll go back and fix the spotted issues via secondary PRs.

**Pull request merged**
